### ISSUE TITLE
Inconsistent masking states and lost settings

### ISSUE THREAD

**Issue by aurelienpierre** [darktable-org/darktable#7372]
For the past 3-4 weeks, I have experienced random issues with masks (darktable master), with and without OpenCL. I don't have a predictable reproducer but they seem to indicate issues with the cache invalidation, threads locks and maybe history stack. They happen with rather heavy history stacks with several drawn masks combined or after enabling/disabling modules several time.

1. Sometimes, the mask preview icon doesn't display the mask, you need to disable/enable the module to get it to work,
2. Sometimes the mask preview shows the whole image while drawn masks are enabled and should narrow the area of the mask. This happens after a couple of enable/disable cycles for a module.
3. Sometimes drawn masks are lost for a module after the pipe recomputes: the GUI shows no mask at all even though several were selected,
4. Once, in the mask manager, after I set a drawn mask to exclusion mode, it made a duplicated mask group appear for the same module and the masks list disappeared in the module widget (so I had to add back all the masks).

I know @ralfbrown has disabled several useless recomputations and such, it could be that a couple of them were needed.

**Comment by ralfbrown** [comment]
Items 1 and 3 definitely sound like a caching issue; item 4 almost certainly is not (that one sounds more like a race condition clobbered internal state).

I've already said a couple of times that the pixelpipe cache needs a full overhaul.  One of the areas where it needs to be improved is in dealing with the fact that the result of a module can change even though the module's parameters haven't (e.g when toggling mask display).  I have been thinking about a redesign of the cache off and on over the past week, and will probably have an RfC with a proposed architecture later this month.

**Comment by lapineige** [comment]
Same issue with version 3.4.0, using clipping warning tool, in luminance only mode: clipping warning on clipped dark areas fails randomly when activated or when the image is changed.

**Comment by todd-prior** [comment]
I was attempting to help someone today understand the masking and I noted that on a simple selection ...tone curve with parametric mask on one of the hue's in the image and then a drawn mask added with partial overlap. I was going to use it to demonstrate channel polarity and inclusive and exclusive, however toggling to inclusive and then back to exclusive does not restore the mask often left with a monochrome image and you need to reset everything. I am noting this on release version 3.4 in WIndows. So it seems that there are problems resetting the mask to the desired combination in a combined mask if you work through some different settings after constructing the initial mask....

**Comment by lapineige** [comment]
I have another bug, I don't know if it's related but I suppose it's the case: if I copy-paste the history from one RAW to another, sometimes it fails, kind of randomly. If I do it again, most of the time it works (sometimes I need to do it 3 times).
Here is the result (I pasted an history stack from another image to these images, one succeeded but not the other):
![cleaned](https://user-images.githubusercontent.com/8353704/103460450-891e0f80-4d16-11eb-802d-f00acb0c91e3.png)

I did have similar rendering issues within the darkroom, using masks (that's not the case for these 2 images) as described above. Going out of the darkroom and entering it again solved the issue.

**Comment by lapineige** [comment]
To elaborate on my last message: I recently found a workaround for that bug, it happens if I copy-paste the history stack from one image to another that was not opened first (imported, but not opened in the darkroom). It works normally if I do : Open the image in darkroom → come back to the lighttable → paste an history stack from another image (using any of the copy buttons).

Previously the only way I had found to copy an history stack was to copy it in several parts, mostly two parts 1) all "corrective" modules, up to chromatic aberration, orientation and so on 2) all the rest. But I couldn't find a particular module that was at the origin of this bug.

And that bug, for similar history stack, doesn't happen all the time for all images… but when it does, not matter which image I'm copying to, I encounter it.

Maybe it's not related to that cache issue ?

(Using Darktable 3.4.1 now)
